### PR TITLE
Added support for correctly sharing process memory regions between threads.

### PR DESCRIPTION
**Pull request by rahahahat**
This PR contains the fix for the arm-clang issue in which threads belonging to the same group we're not sharing process memory regions correctly.

### CODE REVIEW

File: src/lib/OS/MemRegion.cc [hunk 1]
```diff
@@ -123,10 +118,10 @@ uint64_t MemRegion::addVma(VMA* vma, uint64_t startAddr) {
   // same. Here startAddr is either mmap pointer or an address greater than mmap
   // pointer.
   if (!allocated) {
-    startAddr = startAddr >= mmapPtr_ ? startAddr : mmapPtr_;
+    startAddr = startAddr >= mmapReg_->mmapPtr ? startAddr : mmapReg_->mmapPtr;
```
Comment (FinnWilkinson) [inline src/lib/OS/MemRegion.cc:121]: Could you add brackets around `startAddr >= mmapReg_->mmapPtr` for readability please

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]

File: src/lib/OS/SimOS.cc [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #include "simeng/OS/SimOS.hh"
 
 #include <cstdint>
+#include <iomanip>
```
[the remark this answers is not in the document]
Comment (rahahahat) [reply src/lib/OS/SimOS.cc:4]: removed

File: src/include/simeng/OS/MemRegion.hh [hunk 3]
```diff
@@ -34,6 +36,63 @@ struct VMALinkedList {
...
+   * region. */
+  uint64_t stackStart;
+
+  /** Empty constructor for the Stack ProcessRegion. */
+  ProcessStackRegion() : Range(), initialStackPtr(0) {}
+
+  /** Constructor which initialises the ProcessStackRegion with specific
+   * values. Since the stack grows down from higher addresses towards lower
+   * addresses, an additional member variable "stackStart" is used which
+   * represents the same. */
```
[the remark this answers is not in the document]
Comment (rahahahat) [reply src/include/simeng/OS/MemRegion.hh:56]: Removed the extra comment as it was repetitive

File: src/include/simeng/OS/MemRegion.hh [hunk 4]
```diff
@@ -115,46 +176,19 @@ class MemRegion {
...
-
-  /** Upper bound of the process mmap region. */
-  uint64_t mmapEnd_;
-
-  /** Current end address of process mmap region. */
-  uint64_t mmapPtr_;
-
-  /** Size of the mmap region. */
-  size_t mmapSize_;
+  uint64_t memSize_;
```
[the remark this answers is not in the document]
Comment (rahahahat) [reply src/include/simeng/OS/MemRegion.hh:191]: Sorry, the comment is misleading, `memSize` represents the entire size of the process image. I have renamed the variable and reworded the comment as well. Also as far as `getStackStart` and `getProcessImageSize` are concerned both return correct values. `getStackStart` should return `memSize` at it represents the bottom-most address of the entire process image, the bottom-most address of the process image is start of the stack region of the process image.

File: src/lib/OS/MemRegion.cc [hunk 5]
```diff
@@ -13,49 +13,47 @@ MemRegion::MemRegion(uint64_t stackSize, uint64_t heapSize, uint64_t mmapSize,
...
 MemRegion::~MemRegion() {}
 
-uint64_t MemRegion::getStackStart() const { return stackStart_; }
+uint64_t MemRegion::getStackStart() const { return stackReg_.stackStart; }
 
-uint64_t MemRegion::getStackEnd() const { return stackEnd_; }
+uint64_t MemRegion::getStackEnd() const {
+  // Since the stack grows down towards lower addresses, the stack end
+  // address is the lowest address in the stack address range.
+  return stackReg_.start;
```
[the remark this answers is not in the document]
Comment (rahahahat) [reply src/lib/OS/MemRegion.cc:31]: done

File: src/lib/OS/MemRegion.cc [hunk 6]
```diff
@@ -13,49 +13,47 @@ MemRegion::MemRegion(uint64_t stackSize, uint64_t heapSize, uint64_t mmapSize,
...
-uint64_t MemRegion::getMmapStart() const { return mmapStart_; }
+uint64_t MemRegion::getMmapStart() const { return mmapReg_->start; }
 
 uint64_t MemRegion::getMemSize() const { return memSize_; }
 
 uint64_t MemRegion::updateBrkRegion(uint64_t newBrk) {
-  if (newBrk < heapStart_) {
-    return brk_;
+  if (newBrk < heapReg_->start) {
+    return heapReg_->brk;
```
[the remark this answers is not in the document]
Comment (rahahahat) [reply src/lib/OS/MemRegion.cc:54]: done

[the remark this answers is not in the document]
Comment (rahahahat) [reply src/lib/OS/MemRegion.cc:54]: This was a good spot, there was a small inconsistency in the logic. Adding the comment highlighted the problem I have fixed it now.

File: src/include/simeng/OS/MemRegion.hh [hunk 7]
```diff
@@ -42,13 +99,16 @@ class MemRegion {
   /** This constructor creates a MemRegion with values specified by the owning
    * process. */
   MemRegion(uint64_t stackSize, uint64_t heapSize, uint64_t mmapSize,
-            uint64_t memSize, uint64_t stackStart, uint64_t heapStart,
+            uint64_t procImgSize, uint64_t stackStart, uint64_t heapStart,
             uint64_t mmapStart, uint64_t initStackPtr,
             std::function<uint64_t(uint64_t, size_t)> unmapPageTable);
 
   /** This constructor creates am empty MemRegion.*/
```
Comment (jj16791) [inline src/include/simeng/OS/MemRegion.hh:106]: "creates an"

**Review by jj16791 — CHANGES_REQUESTED** [review]

**Review by jj16791 — APPROVED** [review]

**Review by FinnWilkinson — APPROVED** [review]

**Review by FinnWilkinson — APPROVED** [review]

**Pull request merged**